### ISSUE TITLE
Feature request: Gradient for number of loop iterations for automatic differentiation of physics simulations

### ISSUE THREAD

**Issue by AlexHarn** [tensorflow/tensorflow#18920]
## System information
**Not relevant.**

Have I written custom code **yes**
OS Platform and Distribution **different linux systems**
TensorFlow installed from **source and pip** (on different systems)
TensorFlow version **different versions, all >= 1.5.0**
Bazel version **different versions on different systems**
CUDA/cuDNN version **different versions, all >= 8**
GPU model and memory **GTX 1080, GTX 1060, Tesla P40** (potentially many others on our cluster, but not tested yet)
Exact command to reproduce **run [bug_test.py](https://github.com/AlexHarn/tf-ice-model-optimization/blob/1a9f9e11d55277bd595266410bf10a54bbd8b5f0/bug_test.py)**

### Motivation
(you might want to skip this)
Tensorflow is a powerful framework that can potentially be used for much more than static computations like those needed to train neural networks. We (a couple of physicists) are trying to use Tensorflow to get a gradient on some physical parameters to fit those parameters to measured data by propagating the gradient through the entire physics simulation to replicate the data. This is extremely relevant for a lot of people and the idea is not new.  Instead of performing extremely time consuming (in terms of computational and human work time) grid searches on high dimensional parameter spaces we would like to be able to directly perform gradient descent on such spaces in an automated fashion by using Tensorflows automatic differentiation. You can find our work in progress repository [here](https://github.com/AlexHarn/tf-ice-model-optimization).

### The Problem
Like explicitly stated [here](http://download.tensorflow.org/paper/white_paper_tf_control_flow_implementation_2017_11_1.pdf) on page 15 "This means that we assume that `pred` is not trainable", the number of loop iterations is just a constant while backpropagating the gradient. This leads to some extremely unexpected behavior like described in this [issue](https://github.com/AlexHarn/tf-ice-model-optimization/issues/1) (at least unexpected for someone who is a Tensorflow newbie like I am). For our plans to work it is mandatory for the gradient to include information on how the number of loop iterations would have changed, if the trainable variables changed. 

### Example
(You might want to skip this)
In our project we have a trainable variable, which describes the mean distance between scattering events of photons. Tensorflow is able to propagate the gradient through the simulation (while loop), but the gradient only "thinks" that the photons will reach further points when the scattering length is longer, but "is not aware" of the fact that there will be less iterations. 

### Request
I do not know if this is even possible with how Tensorflow works, but we would love to see a way to make this work. Tensorflow would be a really great tool for us to use. So is this something you might work on in the future? 

Also: does anyone maybe have any idea of how to solve or get around this right now with the current implementation? We have put quite some thought into this ourselves already, but until now we have not found a working solution.

## Source code
For a work in progress example of how this would be useful you can take a look at the [master](https://github.com/AlexHarn/tf-ice-model-optimization) and [experimental](https://github.com/AlexHarn/tf-ice-model-optimization/tree/experimental) branches of our repo. 

For a stripped down minimal one dimensional example of the simulation you can take a look at [minimal_1d.py](https://github.com/AlexHarn/tf-ice-model-optimization/blob/gradient-sign-bug/minimal_1d.py), this in its current state however actually converges. For some more information on the unexpected behavior take a look at my [issue](https://github.com/AlexHarn/tf-ice-model-optimization/issues/1) where I reference commits, which demonstrate what's happening.

A minimal example, which does not include any physics and purely demonstrates the unexpected behavior is given [here](https://github.com/AlexHarn/tf-ice-model-optimization/blob/1a9f9e11d55277bd595266410bf10a54bbd8b5f0/bug_test.py).

All of the code includes detailed comments. 

If you are interested in the background and general idea you can take a look at my [slides](https://drive.google.com/file/d/17WdIcpIzFDZmQvD-DCqonmDvz_S0-Arz/view?usp=sharing), even though I am not sure how much of a help those will be for someone from outside. Slides 5 to 8 might be helpful for everyone though. Especially slide 5 which shows how we fit the data using simulations. For the general concept testing right now the data is also given by the same simulation with "true" parameters.

We really hope to get some form of feedback here and I am sorry if this is a trivial or stupid request in the eyes of someone more experienced with Tensorflow.

**Comment by tensorflowbutler** [comment]
Thank you for your post. We noticed you have not filled out the following field in the issue template. Could you update them if they are relevant in your case, or leave them as N/A? Thanks.
Have I written custom code
OS Platform and Distribution
TensorFlow installed from
TensorFlow version
Bazel version
CUDA/cuDNN version
GPU model and memory
Exact command to reproduce

**Comment by AlexHarn** [comment]
I added system information, even though that should not be relevant.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 17 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 32 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 47 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
Nagging Assignee @aselle: It has been 62 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by aselle** [comment]
@alextp, do you think this might be possible with eager? That seems like the best option to me. That allows you to do control flow in python rather than TensorFlow and it does automatic differentiation with a gradient tape (log of operations run in imediate mode). @AlexHarn, does this make sense?

**Comment by alextp** [comment]
I think what you want is very reasonable, but you need different math than plain gradients. Once you do move beyond gradients the many known approaches to this are well-supported by TensorFlow.

The problem lies in (hopefully) continuous functions which are not differentiable everywhere. Anytime you have control flow in a program you have a function which, though it might be continuous, is not guaranteed to have a derivative everywhere. For example, the function f(x) defined by f(x) = x^2 for x < 0, f(x) = x otherwise is continuous, but has no derivative at the change point (0). What it can have is a directional derivative or a subgradient, which is what TensorFlow computes.

If you think about it, by the definition of derivatives (epsilons and deltas, etc) if you make a small change to a number used in a predicate to decide the number of iterations in a loop the number of iterations will not change (for sufficiently small epsilon). There is only a set of measure zero of settings which will change the number of iterations.

So if you want to optimize something which involves the number of iterations you need some other method. A popular one is to define auxiliary losses in your training problem (for example, make the probability of stopping at this loop iteration a function of some parameters, and then add a term to your loss function which maximizes this probability; this will encourage the model to stop earlier). A generalization of this is to write a differentiable upper bound on your number of iterations which you can then optimize. Work like [adaptive computation time rnns](https://arxiv.org/abs/1603.08983) is based on this upper-bound approach.

Another way of dealing with this non-differentiability is to simply pretend things are differentiable and not use a bound. For example, [bengio](https://arxiv.org/abs/1308.3432) shows that in many cases you can approximate the "gradient" of a step function by pretending that the step function isn't there (the staight through estimator).

Finally, approaches such as policy gradient and the REINFORCE algorithm (many sources on the internet; I like [Sergey Levine's slides](http://rail.eecs.berkeley.edu/deeprlcourse-fa17/f17docs/lecture_4_policy_gradient.pdf)) are a general (and so slow) way of optimizing these functions;  you can use them to approximate gradients to change your number of loop iterations.

All of these approaches can be (and have been) implemented in TensorFlow, but none of them is just computing gradients. Because it requires some thought to decide what you should do we remain agnostic.
